docs/topics/addons.rst

```diff
@@ -0,0 +1,145 @@
+developers.
+
+
+Activating and configuring add-ons
+==================================
+
+Add-ons and their configuration live in Scrapy's
+:class:`~scrapy.addons.AddonManager`. During a :class:`~scrapy.crawler.Crawler`
+initialization the add-on manager will read a list of enabled add-ons from your
+``ADDONS`` setting and their optional configuration from the respective
+settings.
+
+The ``ADDONS`` setting is a dict in which every key is an addon class or its
+import path and the value is its priority.
+
+The configuration of an add-on, if necessary at all, is stored as a dictionary
+setting whose name is the uppercase add-on name.
+
+This is an example where two add-ons (in this case with one requiring no
+configuration) are enabled/configured in a project's ``settings.py``::
+
+    ADDONS = {
+        'path.to.someaddon': 0,
+        path.to.someaddon2: 1,
+    }
+
+    SOMEADDON = {
+        'some_config': True,
+    }
+
+
+Writing your own add-ons
+========================
+
+Add-ons are (any) Python *objects* that provide Scrapy's *add-on interface*:
+
+.. attribute:: name
+
+    string with add-on name
+
+    :type: ``str``
+
+.. method:: update_settings(config, settings)
+
+    This method is called during the initialization of the
+    :class:`~scrapy.crawler.Crawler`. Here, you should perform dependency checks
+    (e.g. for external Python libraries) and update the
+    :class:`~scrapy.settings.Settings` object as wished, e.g. enable components
+    for this add-on or set required configuration of other extensions.
+
+    :param config: Configuration of this add-on
+    :type config: ``dict``
+
+    :param settings: The settings object storing Scrapy/component configuration
+    :type settings: :class:`~scrapy.settings.Settings`
+
+.. method:: check_configuration(config, crawler)
+
+    This method is called when the :class:`~scrapy.crawler.Crawler` has been
+    fully initialized, immediately before it starts crawling. You can perform
+    additional dependency and configuration checks here.
+
+    :param config: Configuration of this add-on
+    :type config: ``dict``
+
+    :param crawler: Fully initialized Scrapy crawler
+    :type crawler: :class:`~scrapy.crawler.Crawler`
+
+
+Add-on base class
+=================
+
+Scrapy comes with a built-in base class for add-ons which provides some
+convenience functionality: the add-on configuration can be exposed into
+Scrapy's settings via :meth:`~scrapy.addons.Addon.export_config`, configurable
+via :attr:`~scrapy.addons.Addon.default_config` and
+:attr:`~scrapy.addons.Addon.config_mapping`.
+
+By default, the base add-on class will expose the add-on configuration into
+Scrapy's settings namespace, in upper case. It is
+easy to write your own functionality while still being able to use the
+convenience functions by overwriting
+:meth:`~scrapy.addons.Addon.update_settings`.
+
+.. module:: scrapy.addons
+   :noindex:
+
+.. autoclass:: Addon
+   :members:
```

Comment:
I am not strongly against it, but we could probably do without a base class at all.

Reply:
If we remove the per-addon automatic setting names then it's no longer useful, yeah.

Reply:
> assuming addons are Scrapy components, meaning they can define a from_crawler class method

They are currently not Scrapy components though, we need to add a `create_instance` call to AddonManager for this to work.